src/algorithms/sort/counting-sort.ts

```diff
@@ -0,0 +1,55 @@
+/**
+ * Sort an array of integers in-place and in ascending order using counting sort.
+ *
+ * Time complexity: O(n + k), where n is the number of elements in arr
+ * and k is the range of numbers in arr.
+ *
+ * Space complexity: O(n + k)
+ *
+ * @param arr - array of integers to sort
+ */
+const countingSort = (arr: number[]): void => {
+  if (arr.length == 0) {
+    return
+  }
+
+  // check that arr elements are integers
+  for (let idx = 0; idx < arr.length; idx++) {
+    if (!Number.isInteger(arr[idx])) {
+      console.error(`arr[${idx}] is not an integer`)
+      return
+    }
+  }
+
+  // implementation of counting sort below
+  const max = Math.max(...arr) // find largest integer in arr
+  const min = Math.min(...arr) // find smallest integer in arr
+  const range = max - min + 1 // calculate range of possible numbers
+  const counts: number[] = new Array(range).fill(0) // contains counts for each number in arr
```

Comment:
great comments, ty 🏆

(Comments are usually a sign of codesmell in codebases because the code should explain itself, but they are highly encouraged in this codebase for learning purposes)